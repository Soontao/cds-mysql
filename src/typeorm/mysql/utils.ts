
/**
 * equal string without case sensitive
 * 
 * @param s1 first string
 * @param s2 second string
 * @returns true if both two string are equal without case sensitive
 */
export function equalWithoutCase(s1: string, s2: string) {
  if (s1 === undefined && s2 === undefined) {
    return true;
  }
  if (s1?.length === 0 && s2?.length === 0) {
    return true;
  }
  if(s1 === s2) {
    return true;
  }
  if (s1?.length === s2?.length && s1?.toLowerCase() === s2?.toLowerCase()) {
    return true;
  }

  return false;
}