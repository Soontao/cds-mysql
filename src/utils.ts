
/**
 * overwrite cds compiler type mapping
 */
export function overwriteCDSCoreTypes() {

  const { cdsToSqlTypes } = require("@sap/cds-compiler/lib/render/utils/common");

  cdsToSqlTypes.sqlite = {
    "cds.Binary": "CHAR",
    "cds.hana.BINARY": "CHAR",
    "cds.hana.SMALLDECIMAL": "DECIMAL",
  };

}