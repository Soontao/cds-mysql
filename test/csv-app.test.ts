import { doAfterAll } from "./utils";
import { cwdRequireCDS, setupTest } from "cds-internal-tool";
import path from "path";
import MySQLDatabaseService from "../src";


describe("CSV App Test Suite", () => {

  const client = setupTest(__dirname, "./resources/csv-app");

  afterAll(doAfterAll);

  it("should support deploy by API", async () => {
    const db: MySQLDatabaseService = cwdRequireCDS().db as any;
    await db.deployCSV();
  });

  it("should support automatically migrate csv data", async () => {
    const response = await client.get("/app/$metadata");
    expect(response.status).toBe(200);
    expect(response.data).toMatch(/Peoples/);
    const p1r = await client.get("/app/Peoples(1)");
    expect(p1r.status).toBe(200);
    expect(p1r.data.Name).toBe("Theo 1");
    const { data } = await client.get("/app/Peoples");
    expect(data).toMatchSnapshot();

  });

  it("should be empty when default value input", async () => {
    const { data, status } = await client.get("/app/Areas");
    expect(status).toBe(200);
    expect(data.value).toHaveLength(0);
  });

  it("should match the first record by axios client", async () => {
    const t1r = await client.get("/app/TypeEntity(1)");
    expect(t1r.data).toMatchSnapshot();
  });

  it("should match the first record by cds API", async () => {
    const values = await cds.run(SELECT.one.from("test.resources.csv.app.srv.AppService.TypeEntity", 1));
    expect(values).toMatchSnapshot();
  });

  it("should match the second record by axios client", async () => {
    const { data } = await client.get("/app/TypeEntity(2)");
    expect(data).toMatchSnapshot();
  });

  it("should support ISO format date string", async () => {
    const { data } = await client.get("/app/TypeEntity(3)");
    expect(data).toMatchSnapshot();
  });


  it("should match the second record by cds API", async () => {
    const values = await cds.run(SELECT.one.from("test.resources.csv.app.srv.AppService.TypeEntity", 2));
    expect(values).toMatchSnapshot();
  });

  it("should support search", async () => {
    const res1 = await client.get("/app/Houses?$search=Chengdu");
    expect(res1.status).toBe(200);
    expect(res1.data).toMatchSnapshot();

    const res2 = await client.get("/app/Houses?$search=2357853");
    expect(res2.data).toMatchSnapshot();

  });

  it("should support filter by datetime", async () => {
    const res = await client.get("/app/TypeEntity?$filter=Sign eq 2022-06-13T12:35:10Z");
    expect(res.status).toBe(200);
    expect(res.data?.value?.[0]?.ID).toBe(3);
  });

  it("should support filter by datetimeoffset", async () => {
    const res = await client.get("/app/TypeEntity?$filter=SignTmp eq 2022-06-12T12:35:10.000Z");
    expect(res.status).toBe(200);
    expect(res.data?.value?.[0]?.ID).toBe(3);
  });

  it("should support filter by datetimeoffset with timezone", async () => {
    const res = await client.get("/app/TypeEntity?$filter=SignTmp eq 2022-06-12T20:35:10.000+08:00");
    expect(res.status).toBe(200);
    expect(res.data?.value?.[0]?.ID).toBe(3);
  });


  it("should support migrate again", async () => {
    cwdRequireCDS().db?.["deploy"]?.(
      await cwdRequireCDS().load("*", { root: path.join(__dirname, "./resources/csv-app") })
    );
  });

  for (const aggregation of ["sum", "min", "max", "average", "countdistinct"]) {
    it(`should support simple ${aggregation} aggregation`, async () => {
      const res1 = await client.get(`/app/Houses?$apply=aggregate(price with ${aggregation} as ${aggregation}_price)`);
      expect(res1.status).toBe(200);
      expect(res1.data).toMatchSnapshot();
    });

  }

  it("should support migrate to v2 data", async () => {
    const db: MySQLDatabaseService = cwdRequireCDS().db as any;
    await db.deployCSV(
      undefined,
      [
        "./resources/csv-app/db/__data_v2__/test_resources_csv_app_db-Area.csv",
        "./resources/csv-app/db/__data_v2__/test_resources_csv_app_db-House.csv",
      ].map(p => path.join(__dirname, p))
    );
  });

  it("should not be empty when migrated to v2", async () => {
    const { data, status } = await client.get("/app/Areas");
    expect(status).toBe(200);
    expect(data.value).toMatchSnapshot();
  });

  it("should update house after migrated to v2", async () => {
    const { data, status } = await client.get("/app/Houses(97dc9bc4-f61e-11ec-b939-0242ac120002)");
    expect({ data, status }).toMatchSnapshot();
  });

});
