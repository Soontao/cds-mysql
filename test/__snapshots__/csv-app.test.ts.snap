// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV App Test Suite - Feature (big js enabled) should get error when create duplicated record 1`] = `
{
  "status": 500,
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should match the first record by axios client 1`] = `
{
  "@odata.context": "$metadata#TypeEntity/$entity",
  "Active": true,
  "Age": 19,
  "BirthDay": "1995-11-11",
  "BlobDoc": "SGVsbG8gQ0RTIEJpbmFyeQ==",
  "GlobalUUID": "0da69f3e-3e49-46c6-9eee-3e7e62ddb15d",
  "Height": 177.22,
  "ID": 1,
  "IDCard": 99132132,
  "Name": "Theo 1",
  "Sign": "2022-06-10T12:35:10Z",
  "SignTime": "20:35:10",
  "SignTmp": "2022-06-10T12:35:10.000Z",
  "Weight": 123,
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should match the first record by cds API 1`] = `
{
  "Active": true,
  "Age": 19,
  "BirthDay": "1995-11-11",
  "BlobDoc": {
    "data": [
      72,
      101,
      108,
      108,
      111,
      32,
      67,
      68,
      83,
      32,
      66,
      105,
      110,
      97,
      114,
      121,
    ],
    "type": "Buffer",
  },
  "GlobalUUID": "0da69f3e-3e49-46c6-9eee-3e7e62ddb15d",
  "Height": 177.22,
  "ID": "1",
  "IDCard": "99132132",
  "Name": "Theo 1",
  "Sign": "2022-06-10T12:35:10Z",
  "SignTime": "20:35:10",
  "SignTmp": "2022-06-10T12:35:10.000Z",
  "Weight": "123",
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should match the second record by axios client 1`] = `
{
  "@odata.context": "$metadata#TypeEntity/$entity",
  "Active": false,
  "Age": 23,
  "BirthDay": "1981-05-07",
  "BlobDoc": "d3d3dw==",
  "GlobalUUID": "df1b059a-403d-4dd9-b066-2c9fbd03af7e",
  "Height": 15435.43,
  "ID": 2,
  "IDCard": 3213214312412341,
  "Name": "Theo 2",
  "Sign": "2022-06-10T12:35:10Z",
  "SignTime": "20:35:10",
  "SignTmp": "2022-06-10T12:35:10.000Z",
  "Weight": 322,
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should match the second record by cds API 1`] = `
{
  "Active": false,
  "Age": 23,
  "BirthDay": "1981-05-07",
  "BlobDoc": {
    "data": [
      119,
      119,
      119,
      119,
    ],
    "type": "Buffer",
  },
  "GlobalUUID": "df1b059a-403d-4dd9-b066-2c9fbd03af7e",
  "Height": 15435.43,
  "ID": "2",
  "IDCard": "3213214312412341",
  "Name": "Theo 2",
  "Sign": "2022-06-10T12:35:10Z",
  "SignTime": "20:35:10",
  "SignTmp": "2022-06-10T12:35:10.000Z",
  "Weight": "322",
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should not be empty when migrated to v2 1`] = `
[
  {
    "ID": "5e42cf9c-6fc1-11ed-a1eb-0242ac120002",
    "Name": "Area 1",
    "Size": 10000000.132,
  },
  {
    "ID": "5e42d4ce-6fc1-11ed-a1eb-0242ac120002",
    "Name": "Area 2",
    "Size": 9999433,
  },
]
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support ISO format date string 1`] = `
{
  "@odata.context": "$metadata#TypeEntity/$entity",
  "Active": false,
  "Age": 23,
  "BirthDay": "1981-05-07",
  "BlobDoc": null,
  "GlobalUUID": "df1b059a-403d-4dd9-b066-2c9fbd03af7f",
  "Height": 15435.43,
  "ID": 3,
  "IDCard": 581493248932,
  "Name": "Theo 3",
  "Sign": "2022-06-13T12:35:10Z",
  "SignTime": "20:35:10",
  "SignTmp": "2022-06-12T12:35:10.000Z",
  "Weight": 322,
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support automatically migrate csv data 1`] = `
{
  "@odata.context": "$metadata#Peoples",
  "value": [
    {
      "ID": 1,
      "Name": "Theo 1",
    },
    {
      "ID": 2,
      "Name": "Theo 2",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support search 1`] = `
{
  "@odata.context": "$metadata#Houses",
  "value": [
    {
      "ID": "97dc99ee-f61e-11ec-b939-0242ac120002",
      "address_City": "Chengdu",
      "address_Country": null,
      "address_Province": null,
      "address_Street": null,
      "price": 100000,
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support search 2`] = `
{
  "@odata.context": "$metadata#Houses",
  "value": [
    {
      "ID": "97dc9cb4-f61e-11ec-b939-0242ac120002",
      "address_City": "Shanghai",
      "address_Country": null,
      "address_Province": null,
      "address_Street": null,
      "price": 2357853,
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support simple average aggregation 1`] = `
{
  "@odata.context": "$metadata#Houses(average_price)",
  "value": [
    {
      "@odata.id": null,
      "average_price": 671400,
      "average_price@odata.type": "#Decimal",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support simple countdistinct aggregation 1`] = `
{
  "@odata.context": "$metadata#Houses(countdistinct_price)",
  "value": [
    {
      "@odata.id": null,
      "countdistinct_price": 4,
      "countdistinct_price@odata.type": "#Decimal",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support simple max aggregation 1`] = `
{
  "@odata.context": "$metadata#Houses(max_price)",
  "value": [
    {
      "@odata.id": null,
      "max_price": 2357853,
      "max_price@odata.type": "#Decimal",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support simple min aggregation 1`] = `
{
  "@odata.context": "$metadata#Houses(min_price)",
  "value": [
    {
      "@odata.id": null,
      "min_price": 23423,
      "min_price@odata.type": "#Decimal",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should support simple sum aggregation 1`] = `
{
  "@odata.context": "$metadata#Houses(sum_price)",
  "value": [
    {
      "@odata.id": null,
      "sum_price": 2685600,
      "sum_price@odata.type": "#Decimal",
    },
  ],
}
`;

exports[`CSV App Test Suite - Feature (big js enabled) should update house after migrated to v2 1`] = `
{
  "data": {
    "@odata.context": "$metadata#Houses/$entity",
    "ID": "97dc9bc4-f61e-11ec-b939-0242ac120002",
    "address_City": "Beijing Updated",
    "address_Country": null,
    "address_Province": null,
    "address_Street": null,
    "price": 204323,
  },
  "status": 200,
}
`;
